### ISSUE TITLE
Rich Text editor Control

### ISSUE THREAD

**Issue by russgove** [SharePoint/sp-dev-fx-controls-react#20]
[x ] Enhancement

A rich text editor that lets users 'enhanced' rich text inline, like in classic pages.

**Comment by sebastienlevert** [comment]
Could that be a great starting point? https://github.com/SharePoint/sp-dev-fx-webparts/tree/master/samples/react-textboxio. Franck Cornu did a great job there... Could be interesting to wrap that as a reusable control!

**Comment by estruyf** [comment]
@russgove and @sebastienlevert this is indeed a good starting point. We are looking into using the code of the first party control, but this might take some time. So, we could start by using this one in combination with the third party library dependency. In the future, we might deprecate it, and replace it with the one coming from SharePoint itself.

**Comment by russgove** [comment]
I am crating one for my current project, but it uses CKEditor. I would be glad to contribute  if thats OK.

**Comment by estruyf** [comment]
@russgove that would be great!

**Comment by russgove** [comment]
Just to be clear though, the 'toolbar' only works if your in classic pages.
This is fine for spfx webparts being used as listforms because you cannot
use spfx webparts on listforms in  modern pages. Not a big deal.

**Comment by russgove** [comment]
ok, I moved my  code into my local branch at https://github.com/russgove/sp-dev-fx-controls-react/tree/rte.
Can you check it out there and give me some feedback, or should I do a commit?

**Comment by russgove** [comment]
btw, we really should identifiy all the sharepoint FIELD controls we think we will need (Date, RTE, taxonomy, DateTimePicker, etc...) and come up with some kind of common/reccommended interface to use before we start creating them all. first step would be to identify the controls we think we need.  I'm talking about controls to edit list fields here. Perhaps ListFieldCOntrols should have a particular naming convention as well so users know what its meant for. The  TextFieldWithEdit in my repo is meant to be a generic control. I use it on a page with some text where i want to let the user change some thext and call a function when he saves text, It uses the The RichTextEditor  control.  The RichTextEditor  control  is suuitable for use on a new/editform.

**Comment by estruyf** [comment]
@russgove I have seen it and tested it out. Behaviour still feels the same as before. This is one of the hardest controls to get right, but I feel if we release this, it needs to be providing additional value. ATM the control does not give you any controls to manage text like heading, italic, font weight, ...

**Comment by russgove** [comment]
hey, i'm not sure what you mean by 'behavior still feels the same as before'... this is. a new control. 

yo can adjust fonts using the  controls menu. are you not seeing that after installing?

**Comment by estruyf** [comment]
This is the experience I have when using your branch.

<img width="471" alt="screenshot 2018-02-05 09 03 02" src="https://user-images.githubusercontent.com/2900833/35793830-87d95390-0a53-11e8-84ca-d4587976fc8e.png">

**Comment by russgove** [comment]
yeah, thats what I mentioned earlier... Just to be clear though, the 'toolbar' only works if you're in classic pages.

I was just looking at tinymce editor. The api is almost identical. When I get sone time i'll  switch to tinymce and see if that works better.

**Comment by estruyf** [comment]
Don't invest to much time, because at some point we might be able to switch to the one from SharePoint.

**Comment by hbgeorgian** [comment]
I duplicated the one from SharePoint for my projects, using textbox.io. I'd be happy to share what I've done.
![custom textbox](https://user-images.githubusercontent.com/16600066/37127201-b19d7c5e-2242-11e8-83b9-e7cdda22d488.png)

Let me how I can contribute!

**Comment by ErnstWolthaus** [comment]
@hbgeorgian any change of sharing the code?

**Comment by NigelWit** [comment]
Hey Elio, did this control get merged?

@hbgeorgian - any chance of submitting this control?

**Comment by msgx** [comment]
@hbgeorgian, you should submit a [pull request](https://help.github.com/articles/about-pull-requests/) (PR) as described here: https://help.github.com/articles/creating-a-pull-request/

**Comment by estruyf** [comment]
The problem of this would be the license of textbox.io, check this issue: https://github.com/SharePoint/sp-dev-fx-webparts/issues/410 - So we cannot get this included in the library.

**Comment by hugoabernier** [comment]
Would the same control, but created using [Quill](https://quilljs.com) work? I'd be happy to take this one on.

**Comment by hugoabernier** [comment]
I submitted a PR for a richtext control a month ago. 

Here is what it looks like:
![image](https://github.com/SharePoint/sp-dev-fx-controls-react/raw/c36d7f896aed3ba557f150158fe6c55afcf9317a/docs/documentation/docs/assets/RichTextOutput.gif) 

It uses Quill (which is open source) and shouldn't have the same licensing issues as CKE.

Let me know if there is anything I'm missing? (I did submit the changes unsolicited) Not sure if I followed the right process.

**Comment by sebastienlevert** [comment]
That's an amazing implementation! We'll done! Borderline confusing with the OOTB Text webpart haha!

**Comment by jonathanhotono** [comment]
@hugoabernier that looks good! Is there any way to extend to use tables and embed video/image?

**Comment by hugoabernier** [comment]
Thanks! Quill.js (which I use for this implementation) doesn't support tables without additional plug-ins. I didn't want to go through the effort unless there was demand for it.

As for embedding images/videos, I considered it, but wanted to stay as close as possible to the out-of-the-box RTE. Again, if there is demand, I'd be happy to add it.

Let's wait and see if this pull request gets approved first?

**Comment by jonathanhotono** [comment]
Cool. Let's see whether overtime tables and embeds required for editors as we can direct them to use normal OOTB modern webparts to achieve it. 

Could be tricky to implement images as we will need the file explorer component (which microsoft does not share) so user can easily find their images compared to passing normal url.

**Comment by hugoabernier** [comment]
I just submitted a [PR for a sample](https://github.com/SharePoint/sp-dev-fx-webparts/pull/768) in the sp-dev-fx-webparts repo that implements a custom file picker.
![File Picker](https://github.com/SharePoint/sp-dev-fx-webparts/raw/e05ad22090a8f38261839eae34322ca5043bfa19/samples/react-comparer/assets/ComparerWebPart.gif)

Right now, I only tested it for images, but I wrote it with the intent to support other file types.

Maybe we could extract the picker code out of the sample and make it a PnP property control?

**Comment by estruyf** [comment]
@hugoabernier will try to process it later this week, but just found out the MS has its own rich-text editor called Rooster: https://github.com/Microsoft/roosterjs - They also provide a React version: https://github.com/Microsoft/roosterjs-react

Might be better to test this one out first and maybe see if this could replace Quil.js

**Comment by hugoabernier** [comment]
@estruyf Thanks for finding roosterjs!

Let's hold off on processing the PR then, and let me see what I can do about roosterjs and amend/resubmit PR.

**Comment by estruyf** [comment]
@hugoabernier I'll try to figure out how we can solve this, but thank you for testing it out!

**Comment by araver** [comment]
If you don't really care about the warning, I was able to get around the warnings by adding a suppression to gulpfile.js
```javascript
let regex = new RegExp("^(\\(Emitted value instead of an instance of Error\\))*");
build.addSuppression(regex);
```

**Comment by estruyf** [comment]
Just created my own POC for the `roosterjs`, seems that they didn't include their source files to which the sourcemaps are pointing to. That is why you see all these warnings pop-up.

Like @araver mentioned, we could suppress these warnings for now.

**Comment by estruyf** [comment]
Created a new branch with the POC for the `RichtTextEditor` control: https://github.com/SharePoint/sp-dev-fx-controls-react/tree/rooster-poc

**Comment by glyn-clough-wm-reply** [comment]
Any update on this guys? @hugoabernier the screenshot above looks incredible and would be great to have that added to the library. It's exactly what we need!

Out of interest @estruyf if MS use Quil.js for the OOB text editor - even though they have Rooster - then why wouldn't you use that?

**Comment by hugoabernier** [comment]
@glyn-clough-WM-reply I believe the current pr is on hold because I'm rewriting the control to use Rooster.

Rooster is an awesome rte but it definitely has its challenges.  I hope to submit a new pr in the next few days.

**Comment by estruyf** [comment]
The PR for this has been merged and control will be available in `1.13.0`. If you already want to test out the control, you can do that by installing the beta version: https://sharepoint.github.io/sp-dev-fx-controls-react/beta/

**Comment by hugoabernier** [comment]
> The PR for this has been merged and control will be available in 1.13.0. If you already want to test out the control, you can do that by installing the beta version: https://sharepoint.github.io/sp-dev-fx-controls-react/beta/

Thanks @estruyf it'll give me time to reach out to the rooster team and see if we can solve some issues with the custom toolbar support. I haven't given up yet!

**Comment by vermaraj83** [comment]
Hi @estruyf , 
could you please let us know that when RTE control will be available to use in SPFx using react?

**Comment by estruyf** [comment]
The new version is now building and will be released in a couple of minutes.

**Comment by subhasishmukhopadhyay** [comment]
I have created a RichText control to use in List Form customization using custom spfx webpart. It is working fine to save rich text value while creating a new item but when trying to open existing list item then all other fields are loaded properly in the web part but rich text value is showing blank. I am trying to load the values from componetdidmount  using the below code
sp.web.lists.getByTitle("list title").items.getById(<parameter coming from URL>).get().then((item: Item) => {
this.setState({Title: item["Title"]});
        this.setState({Date: new Date(parseInt(item["Date"].split("T")[0].split("-")[0]),parseInt(item["Date"].split("T")[0].split("-")[1])-1,parseInt(item["Date"].split("T")[0].split("-")[2]))});
        this.setState({Month: {key:item["Month"]}});
        this.setState({Year: {key:item["Year"]}});
        this.setState({DocumentType: {key:item["Document_x0020_Type"]}});
        this.setState({BA: {key:item["BA"]}});
        this.setState({Division: {key:item["DIV"]}});
        this.setState({ValueAmount: item["Value"]});
        this.setState({Currency: {key:item["Currency"]}});
        this.setState({DocID: item["Document_x0020_ID"]});
        this.setState({Description: item["Description"]});
});

Here is the rich text control
<RichText value={this.state.Description} onChange={(text) => this.handleDescription(text)} />

Please suggest.

**Comment by sunnysinha05** [comment]
HI Team,

I am also facing the above issue and struggling alot from few days...please update about the issue

**Comment by siddharth-vaghasia** [comment]
@sunnysinha05 .... looks like a long, old and closed thread... it would be better if you open a new issue, which will get require attention and being specific what issue are you facing with a screenshot would help.

**Comment by sunnysinha05** [comment]
@siddharth-vaghasia  @VesaJuvonen 
Already raised this issue with all info  and screenshot, but not getting any updates on it :(
https://github.com/pnp/sp-dev-fx-controls-react/issues/588

**Comment by joelfmrodrigues** [comment]
@lukju yes, it uses quill

**Comment by lukju** [comment]
thanks, @joelfmrodrigues . do you know why the version based on rooster didn't find its way into the product? The dicussion above looks like the POC was successful...

**Comment by joelfmrodrigues** [comment]
@lukju Sorry no idea, I was not involved in the discussion and was not a maintainer of the repo at the time. I will try to reach out to Hugo Bernier and see if I can get more info.

**Comment by joelfmrodrigues** [comment]
@lukju  I think the best idea would actually be to leave this closed thread and open a new issue if there is something that is not working or a suggestion for improvement. We are also happy to review PRs if you want to contribute :)
This thread involved many people and would be challenging to get to the bottom of it

**Comment by TapanKumarBarik** [comment]
Does it have any support like people picker control ?
